server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -63,18 +224,98 @@ public QueryGroupStats nodeStats() {
+                    final double threshold = getNormalisedRejectionThreshold(
+                        queryGroup.getResourceLimits().get(resourceType),
+                        resourceType
+                    );
+                    final double lastRecordedUsage = queryGroupState.getResourceState().get(resourceType).getLastRecordedUsage();
+                    if (threshold < lastRecordedUsage) {
+                        reject = true;
+                        reason.append(resourceType)
+                            .append(" limit is breaching for ENFORCED type QueryGroup: (")
+                            .append(threshold)
+                            .append(" < ")
+                            .append(lastRecordedUsage)
+                            .append("). ");
+                        queryGroupState.getResourceState().get(resourceType).rejections.inc();
+                        // should not double count even if both the resource limits are breaching
+                        break;
+                    }
+                }
+            }
+            if (reject) {
+                queryGroupState.totalRejections.inc();
+                throw new OpenSearchRejectedExecutionException(
+                    "QueryGroup " + queryGroupId + " is already contended. " + reason.toString()
+                );
+            }
+        });
+    }
+
+    private double getNormalisedRejectionThreshold(double limit, ResourceType resourceType) {
+        if (resourceType == ResourceType.CPU) {
+            return limit * workloadManagementSettings.getNodeLevelCpuRejectionThreshold();
+        } else if (resourceType == ResourceType.MEMORY) {
+            return limit * workloadManagementSettings.getNodeLevelMemoryRejectionThreshold();
+        }
+        throw new IllegalArgumentException(resourceType + " is not supported in WLM yet");
+    }
+
+    public Set<QueryGroup> getActiveQueryGroups() {
+        return activeQueryGroups;
+    }
+
+    public Set<QueryGroup> getDeletedQueryGroups() {
+        return deletedQueryGroups;
+    }
+
+    @Override
+    public void onTaskCompleted(Task task) {
+        if (!(task instanceof QueryGroupTask)) {
+            return;
+        }
+        final QueryGroupTask queryGroupTask = (QueryGroupTask) task;
+        String queryGroupId = queryGroupTask.getQueryGroupId();
+
+        // set the default queryGroupId if not existing in the active query groups
+        String finalQueryGroupId = queryGroupId;
+        boolean exists = activeQueryGroups.stream().anyMatch(queryGroup -> queryGroup.get_id().equals(finalQueryGroupId));
+
+        if (!exists) {
+            queryGroupId = QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get();
+        }
+
+        if (task instanceof SearchShardTask) {
+            queryGroupStateMap.get(queryGroupId).shardCompletions.inc();
+        } else {
+            queryGroupStateMap.get(queryGroupId).completions.inc();
```

Comment:
How do we have `MultiSearch` use case? Is that completely independent, without any footprint in the metrics outside of the individual child search requests?

Reply:
All of the individual search requests will still have the same queryGroupId from the MultiSearchRequest, So it will be tracked under same queryGroup.